src/mame/bfm/bfm_blackbox.cpp

```diff
@@ -0,0 +1,1851 @@
+
+static INPUT_PORTS_START( bb_fiest )
+	PORT_INCLUDE( bb_firec )
+	PORT_INCLUDE( coin_10p_2p_50p )
+
+	PORT_MODIFY("IN2000")
+	PORT_DIPNAME( 0x08, 0x08, "Jackpot amount" )
+	PORT_DIPSETTING(    0x00, u8"£2" )
+	PORT_DIPSETTING(    0x08, u8"£1" )
+
+	PORT_MODIFY("PERCENTAGE")
+	PORT_DIPNAME( 0x03, 0x01, "Hold/nudge chance" )
+	PORT_DIPSETTING(    0x00, "Low (21.9%/6.25%)" )
+	PORT_DIPSETTING(    0x01, "High (26.6%/6.25%)" )
+INPUT_PORTS_END
+
+static INPUT_PORTS_START( bb_oal )
+	PORT_INCLUDE( blackbox_inputs_em_opto )
+	PORT_INCLUDE( coin_10p_10pt_50p )
+
+	PORT_MODIFY("IN1800_1")
+	PORT_BIT( 0x04, IP_ACTIVE_HIGH, IPT_START1 )
+	PORT_BIT( 0x08, IP_ACTIVE_HIGH, IPT_BUTTON1 ) PORT_NAME("Gamble")
+	PORT_BIT( 0x10, IP_ACTIVE_HIGH, IPT_POKER_HOLD1 )
+	PORT_BIT( 0x20, IP_ACTIVE_HIGH, IPT_POKER_HOLD2 )
+	PORT_BIT( 0x40, IP_ACTIVE_HIGH, IPT_POKER_HOLD3 )
+	PORT_BIT( 0x80, IP_ACTIVE_HIGH, IPT_GAMBLE_PAYOUT ) PORT_NAME("Collect")
+
+	PORT_MODIFY("IN1800_2")
+	PORT_BIT( 0x10, IP_ACTIVE_HIGH, IPT_BUTTON2 ) PORT_NAME("Nudge Up 1")
+	PORT_BIT( 0x20, IP_ACTIVE_HIGH, IPT_BUTTON3 ) PORT_NAME("Nudge Up 2")
+	PORT_BIT( 0x40, IP_ACTIVE_HIGH, IPT_BUTTON4 ) PORT_NAME("Nudge Up 3")
+
+	PORT_MODIFY("PERCENTAGE")
+	PORT_DIPNAME( 0x03, 0x02, "Hold/nudge chance" )
+	PORT_DIPSETTING(    0x00, "Low (31.2%/6.25%)" )
+	PORT_DIPSETTING(    0x01, "High (31.2%/7.8%)" )
+	PORT_DIPSETTING(    0x02, "Medium (15.6%/7.8%)" )
+INPUT_PORTS_END
+
+static INPUT_PORTS_START( bb_gspin )
+	PORT_INCLUDE( blackbox_inputs_em_opto )
+	PORT_INCLUDE( coin_10p_50p_20p_club )
+
+	PORT_MODIFY("IN1800_2")
+	PORT_BIT( 0x01, IP_ACTIVE_HIGH, IPT_POKER_HOLD1 )
+	PORT_BIT( 0x02, IP_ACTIVE_HIGH, IPT_POKER_HOLD2 )
+	PORT_BIT( 0x04, IP_ACTIVE_HIGH, IPT_POKER_HOLD3 )
+	PORT_BIT( 0x08, IP_ACTIVE_HIGH, IPT_POKER_HOLD4 )
+	PORT_BIT( 0x10, IP_ACTIVE_HIGH, IPT_BUTTON1 ) PORT_NAME("Gamble")
+	PORT_BIT( 0x20, IP_ACTIVE_HIGH, IPT_GAMBLE_TAKE ) PORT_NAME("Collect")
+	PORT_BIT( 0x40, IP_ACTIVE_HIGH, IPT_GAMBLE_HALF ) PORT_NAME("Gamble Half")
+	PORT_BIT( 0x80, IP_ACTIVE_HIGH, IPT_START1 )
+
+	PORT_MODIFY("IN2000")
+	PORT_BIT( 0x10, IP_ACTIVE_HIGH, IPT_POKER_CANCEL )
+	PORT_BIT( 0x20, IP_ACTIVE_HIGH, IPT_INTERLOCK ) PORT_NAME("Back Door") PORT_TOGGLE
+
+	PORT_MODIFY("IN1800_4")
+	PORT_DIPNAME( 0x01, 0x00, "Test Switch 2" ) // Alternate test switches
+	PORT_DIPSETTING(    0x00, "Run" )
+	PORT_DIPSETTING(    0x01, "Test" )
+	PORT_DIPNAME( 0x02, 0x00, "Test Switch 1" )
+	PORT_DIPSETTING(    0x00, "Run" )
+	PORT_DIPSETTING(    0x02, "Test" )
+	PORT_MODIFY("IN1800_5")
+	PORT_BIT( 0x01, IP_ACTIVE_HIGH, IPT_UNUSED )
+	PORT_BIT( 0x02, IP_ACTIVE_HIGH, IPT_UNUSED )
+
+	PORT_MODIFY("PERCENTAGE")
+	PORT_DIPNAME( 0x03, 0x01, "Percentage" )
+	PORT_DIPSETTING(    0x00, "75%" )
+	PORT_DIPSETTING(    0x01, "80%" )
+INPUT_PORTS_END
+
+void blackbox_base_state::machine_start()
+{
+	m_lamps.resolve();
+	m_digits.resolve();
+	m_test_led.resolve();
+
+	save_item(NAME(m_input_en));
+
+	for(int i = 0; i < 6; i++) m_input_en[i] = false;
```

Comment:
Please use a range-based `for` loop or `std::fill`/`std::fill_n` in a way that doesn’t need the array size as a literal here.